checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java

```diff
@@ -1,176 +1,777 @@
       }
 
-      propagateRegPaths(current, visited, worklist);
+      addNonExceptionalSuccessorsToWorklist(obligations, current.block, visited, worklist);
+    }
+
+    addMemberAndClassAnnotations();
+  }
+
+  /**
+   * Adds inferred {@literal @Owning} annotations to fields, {@literal @EnsuresCalledMethods}
+   * annotations to the current method, and {@literal @InheritableMustCall} annotation to the
+   * enclosing class.
+   */
+  private void addMemberAndClassAnnotations() {
+    WholeProgramInference wpi = resourceLeakAtf.getWholeProgramInference();
+    assert wpi != null : "MustCallInference is running without WPI.";
+    for (VariableElement fieldElt : updateOwningFields()) {
+      wpi.addFieldDeclarationAnnotation(fieldElt, OWNING);
+    }
+    if (!disposedFields.isEmpty()) {
+      addEnsuresCalledMethods();
+    }
+
+    addOrUpdateClassMustCall();
+  }
+
+  /**
+   * Returns a set of obligations representing the formal parameters of the current method that have
+   * non-empty MustCall annotations. Returns an empty set if the given CFG doesn't correspond to a
+   * method body.
+   *
+   * @param cfg the control flow graph of the method to check
+   * @return a set of obligations representing the parameters with non-empty MustCall
+   */
+  private Set<Obligation> getNonEmptyMCParams(ControlFlowGraph cfg) {
+    // TODO what about lambdas?
+    if (cfg.getUnderlyingAST().getKind() != UnderlyingAST.Kind.METHOD) {
+      return Collections.emptySet();
+    }
+    Set<Obligation> result = null;
+    for (VariableTree param : methodTree.getParameters()) {
+      if (resourceLeakAtf.declaredTypeHasMustCall(param)) {
+        VariableElement paramElement = TreeUtils.elementFromDeclaration(param);
+        if (result == null) {
+          result = new HashSet<>(2);
+        }
+        result.add(
+            new Obligation(
+                ImmutableSet.of(new ResourceAlias(new LocalVariable(paramElement), param))));
+      }
+    }
+    return result != null ? result : Collections.emptySet();
+  }
+
+  /**
+   * Updates the owning fields set for this class to include all fields inferred as owning in this
+   * iteration.
+   *
+   * @return the owning fields, including fields inferred as owning from the current iteration
+   */
+  private Set<VariableElement> updateOwningFields() {
+    owningFields.addAll(disposedFields);
+    return owningFields;
+  }
+
+  /**
+   * Adds an owning annotation to the formal parameter at the given index.
+   *
+   * @param index the index of a formal parameter of the current method (1-based)
+   */
+  private void addOwningToParam(int index) {
+    WholeProgramInference wpi = resourceLeakAtf.getWholeProgramInference();
+    wpi.addDeclarationAnnotationToFormalParameter(methodElt, index, OWNING);
+  }
+
+  /**
+   * Adds the node to the disposedFields set if it is a field and its must-call obligation is
+   * satisfied by the given method call. If so, it will be given an @Owning annotation later.
```

Comment:
I think if we want to be consistent with the algorithm, when we see somewhere in the code, must-call obligation of a field is satisfied, then the field has to be owning. Since disposedFields does not monotonically increase, I think we need to add it to the owningField set too.

Reply:
> I think we need to add it to the owningField set too

If you're saying that there is a bug in the implementation (which I think you are), then we should write a test case for this situation.

Reply:
I was unable to reproduce a test case because the test infrastructure for WPI can only run for one iteration. I suspect that this issue arises when the @ECM annotation on the finalizer method is inferred during the second iteration.

```
{
  Foo f;

  void overwriteFoo() { 
     f.a();
     f = new Foo();
  }

  finalizer() {
     closeFoo(f);
  }
}
```

Reply:
> I was unable to reproduce a test case because the test infrastructure for WPI can only run for one iteration

This should not prevent you from writing a test case: if you want to test what happens on the second iteration, the input should have the annotations produced by the first iteration. This process works inductively, so it is possible to test WPI behavior on any iteration using the one-iteration test infrastructure.

Reply:
I wasn't able to reproduce this issue because in the test infrastructure, the final result (after inference) should be type checked. If a class has an owning field, it also needs to have a finalizer method `m` that disposes of the field and an `@InheritableMustCall('m')` annotation on the class declaration. This contradicts the scenario I described

Reply:
> I wasn't able to reproduce this issue because in the test infrastructure, the final result (after inference) should be type checked

It's possible to avoid this limitation in one of the following ways:
* use an `@SuppressWarnings` annotation to ignore the other warning that you don't care about testing
* use the same mechanism in the `build.gradle` file that is used by the `ExpectedErrors.java` test case for the `AinferTestChecker`: effectively, disable the removal of the expected warnings.

Of these two, I think in this situation, you should strongly prefer the former.

If you're not able to get this working, rather than a text-only response, can you include the full test case in your next comment on this thread so that I can try to get it working myself?